### PR TITLE
SAMZA-1429: Add callback success/failure metrics to async tasks

### PR DESCRIPTION
**Pull request by dnishimura**
@jmakes & @nickpan47 please take a look when you get the chance.

### CODE REVIEW

File: samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java [hunk 1]
```diff
@@ -91,6 +94,8 @@ public AsyncRunLoop(Map<TaskName, TaskInstance> taskInstances,
     this.threadPool = threadPool;
     this.consumerMultiplexer = consumerMultiplexer;
     this.containerMetrics = containerMetrics;
+    this.containerMetricsCompletedCounter = containerMetrics.newCounter("async-callback-complete-calls");
```
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java]: Please follow the pattern we had to declare/add metric in SamzaContainerMetrics (i.e. instantiate the metric inside SamzaContainerMetrics class and just refer to it from outside like containerMetrics.callbackCompletes.inc()

File: samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java [hunk 2]
```diff
@@ -595,10 +604,14 @@ public void onFailure(TaskCallback callback, Throwable t) {
     private final Set<SystemStreamPartition> processingSspSet;
     private final TaskName taskName;
     private final TaskInstanceMetrics taskMetrics;
+    private final Counter taskMetricsCompletedCounter;
```
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java]: Same here. Please follow the current pattern of instantiating metric in TaskInstanceMetrics

File: samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java [hunk 3]
```diff
@@ -200,6 +201,49 @@ public void setup() {
     when(consumerMultiplexer.pollIntervalMs()).thenReturn(10);
   }
 
+  @Test
```
Comment (nickpan47) [inline samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java]: Can we also add the callback failure test cases as well?

**Comment by dnishimura** [comment]
@jmakes I'll defer to @nickpan47, but it looks like you're correct that `messages-in-flight` will keep increasing if the user never calls the callback methods. Also good point on 2.

**Comment by nickpan47** [comment]
@jmakes from the incidence that I saw, messages-in-flight stays as a constant positive number due to the fact that the process stalls when this number equals task.max.concurrency, which does not help much since in a full-throttle process, messages-in-flight may be a constant as well.

Good point on 2! I missed that. But I think that callbacks-completed metrics may still be helpful in the following situation (i.e. the one I caught previously during debugging some user code):
# task.callback.timeout.ms is not set
# user missed the call TaskCallback.complete() on completion
Mainly, if the container stuck w/ process-calls = 0 and callback-completed = 0 and messages-in-flight != 0, we can conclude that the container got stuck due to callback.complete() is never triggered. Make sense?

**Comment by jmakes** [comment]
@nickpan47 

> user missed the call TaskCallback.complete() on completion
Mainly, if the container stuck w/ process-calls = 0 and callback-completed = 0 and messages-in-flight != 0, we can conclude that the container got stuck due to callback.complete() is never triggered. Make sense?

Well, process-calls is incremented before invoking the user process(), so I don't think that scenario can occur, exactly. Instead process-calls = 1, and messages-in-flight = 1. 

To me, if process-calls is not increasing and messages-in-flight stays non-zero for the same duration, we can conclude that it's stuck, right?

**Comment by nickpan47** [comment]
Synced up with @jmakes offline. We agreed that this task level callback-completed would be useful to debug that a certain task stuck w/ failing. Hence, we are only adding this one metric in TaskInstanceMetrics and removing the failure counts.

File: samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java (hunk 1 above)
Comment (dnishimura) [reply samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java]: Done

File: samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java (hunk 2 above)
Comment (dnishimura) [reply samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java]: Done!

**Comment by dnishimura** [comment]
@nickpan47 I removed the async container level metrics and left only the async completed metric at the task level per your discussion with @jmakes. Ready for another review!

File: samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java [hunk 4]
```diff
@@ -200,6 +201,45 @@ public void setup() {
...
+
+    TestTask task0 = new TestTask(true, true, false, task0ProcessedMessages);
+    TestTask task1 = new TestTask(true, true, false, task1ProcessedMessages);
+    TaskInstance t0 = createTaskInstance(task0, taskName0, ssp0);
+    TaskInstance t1 = createTaskInstance(task1, taskName1, ssp1);
+
+    Map<TaskName, TaskInstance> tasks = new HashMap<>();
+    tasks.put(taskName0, t0);
+    tasks.put(taskName1, t1);
+    //task0.callbackHandler = buildOutofOrderCallback(task0);
```
Comment (nickpan47) [inline samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java]: Remove this line if not used.

File: samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java [hunk 5]
```diff
@@ -200,6 +201,45 @@ public void setup() {
     when(consumerMultiplexer.pollIntervalMs()).thenReturn(10);
   }
 
+  @Test
```
Comment (nickpan47) [inline samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java]: Would be nice if we also add a test case to illustrate that callback-completed was not incremented by mal-implemented task code.

**Pull request closed without merging**